### PR TITLE
Issue 3829 - merged general container/federation endpoints

### LINKED ISSUE #3829: Merge route implementions : general model functions
Similar to https://github.com/3drepo/3drepo.io/issues/3720, but for general model functions (add/delete, model settings etc)

### PR DESCRIPTION
**Pull request by carmenfan**
This fixes #3829

#### Description
- merged general container/federation functions
- fixed some bugs identified

#### Test cases
- functions should work as before
- swagger should make sense

### CODE REVIEW

File: backend/tests/v5/unit/middleware/dataConverter/outputs/teamspaces/projects/models/commons/modelSettings.test.js [hunk 1]
```diff
@@ -80,11 +87,50 @@ const testFormatModelSettings = () => {
...
+			lastUpdated: new Date(),
+			latestRevision: generateUUID(),
+		},
+		errorReason: { timestamp: new Date() } }, 'data to convert'],
+	])('Format model stats data', (isFed, data, desc) => {
+		test(`[${isFed ? 'Federation' : 'Container'}] should format correctly with ${desc}`,
+			async () => {
+				const nextIdx = respondFn.mock.calls.length;
+				await ModelSettingsOutputMiddlewares.formatModelStats(isFed)({ outputData: cloneDeep(data) }, {});
+				expect(respondFn.mock.calls.length).toBe(nextIdx + 1);
```
Comment (ChristosTsiotsias) [inline backend/tests/v5/unit/middleware/dataConverter/outputs/teamspaces/projects/models/commons/modelSettings.test.js:110]: you could use toHaveBeenCalledTimes

File: backend/tests/v5/unit/middleware/dataConverter/outputs/teamspaces/projects/models/commons/modelSettings.test.js [hunk 2]
```diff
@@ -80,11 +87,50 @@ const testFormatModelSettings = () => {
...
+				expect(respondFn.mock.calls[nextIdx][2]).toEqual(templates.ok);
```
Comment (ChristosTsiotsias) [inline backend/tests/v5/unit/middleware/dataConverter/outputs/teamspaces/projects/models/commons/modelSettings.test.js:111]: you could use toHaveBeenCalledWith

File: backend/tests/v5/e2e/routes/teamspaces/projects/models/common/general.test.js [hunk 3]
```diff
@@ -0,0 +1,759 @@
...
+				[`the ${modelType} list contains ${modelType} that is already a favourite`, getRoute(), true, undefined, { [`${modelType}s`]: [favModel._id] }],
+			];
+		};
+
+		const runTest = (desc, route, success, expectedOutput, payload) => {
+			test(`should ${success ? 'succeed' : `fail with ${expectedOutput.code}`} if ${desc}`, async () => {
+				const expectedStatus = success ? templates.ok.status : expectedOutput.status;
+				const res = await agent.patch(route).send(payload).expect(expectedStatus);
+				if (!success) {
+					expect(res.body.code).toEqual(expectedOutput.code);
```
Comment (ChristosTsiotsias) [inline backend/tests/v5/e2e/routes/teamspaces/projects/models/common/general.test.js:367]: we could also check that the model is indeed added to the favourites

File: backend/tests/v5/e2e/routes/teamspaces/projects/models/common/general.test.js [hunk 4]
```diff
@@ -0,0 +1,759 @@
...
+
+			const getRoute = ({
+				projectId = project.id,
+				key = users.tsAdmin.apiKey,
+				modelIds = [model._id],
+			} = {}) => {
+				let url = `/v5/teamspaces/${teamspace}/projects/${projectId}/${modelType}s/favourites`;
+				let firstParam = true;
+				if (key) {
+					url += `${firstParam ? '?' : '&'}key=${key}`;
```
Comment (ChristosTsiotsias) [inline backend/tests/v5/e2e/routes/teamspaces/projects/models/common/general.test.js:408]: firstParam will always be true so maybe we can simplify that

File: backend/tests/v5/e2e/routes/teamspaces/projects/models/common/general.test.js [hunk 5]
```diff
@@ -0,0 +1,759 @@
...
+				[`the ${modelType} list contains correct data`, getRoute({ modelIds: [favModel._id] }), true],
+			];
+		};
+
+		const runTest = (desc, route, success, expectedOutput) => {
+			test(`should ${success ? 'succeed' : `fail with ${expectedOutput.code}`} if ${desc}`, async () => {
+				const expectedStatus = success ? templates.ok.status : expectedOutput.status;
+				const res = await agent.delete(route).expect(expectedStatus);
+				if (!success) {
+					expect(res.body.code).toEqual(expectedOutput.code);
```
Comment (ChristosTsiotsias) [inline backend/tests/v5/e2e/routes/teamspaces/projects/models/common/general.test.js:434]: we could verify that the models are no longer favourites

File: backend/tests/v5/e2e/routes/teamspaces/projects/models/common/general.test.js [hunk 6]
```diff
@@ -0,0 +1,759 @@
...
+				}
+			});
+		};
+		describe.each(generateTestData(true))('Federations', runTest);
+		describe.each(generateTestData())('Containers', runTest);
+	});
+};
+
+describe('E2E routes/teamspaces/projects/federations', () => {
```
Comment (ChristosTsiotsias) [inline backend/tests/v5/e2e/routes/teamspaces/projects/models/common/general.test.js:745]: the name should be updated

**Review by ChristosTsiotsias — CHANGES_REQUESTED** [review]

File: backend/src/v5/middleware/dataConverter/inputs/teamspaces/projects/models/containers.js [hunk 7]
```diff
@@ -29,7 +29,8 @@ Containers.validateUpdateSettingsData = validateUpdateSettingsData(false);
 Containers.canDeleteContainer = async (req, res, next) => {
 	try {
 		const { teamspace, container } = req.params;
-		const fed = await getModelByQuery(teamspace, { 'subModels.model': container }, { _id: 1, name: 1 }).catch(() => {});
+		const query = { $or: [{ 'subModels.model': container }, { subModels: container }] };
```
[the remark this answers is not in the document]
Comment (carmenfan) [reply backend/src/v5/middleware/dataConverter/inputs/teamspaces/projects/models/containers.js:32]: yes, @charence is addressing it in #3845 and he should be removing these workarounds there 😋

File: backend/src/v5/middleware/dataConverter/outputs/teamspaces/projects/models/commons/modelSettings.js [hunk 8]
```diff
@@ -41,4 +42,21 @@ ModelSettings.formatModelSettings = (req, res) => {
 	respond(req, res, templates.ok, formattedSettings);
 };
 
+ModelSettings.formatModelStats = (isFed) => (req, res) => {
+	const { outputData } = req;
+
+	if (isFed) {
+		if (outputData.lastUpdated) outputData.lastUpdated = outputData.lastUpdated.getTime();
```
[the remark this answers is not in the document]
Comment (carmenfan) [reply backend/src/v5/middleware/dataConverter/outputs/teamspaces/projects/models/commons/modelSettings.js:49]: we could, but this literally does the same thing other than it wouldn't stick an `undefined` on a field that doesn't exist, which I personally don't really like

Why do you prefer the other way?

**Review by ChristosTsiotsias — APPROVED** [review]

**Pull request merged**